// @udecode/plate-find-replace/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  MARK_SEARCH_HIGHLIGHT: () => MARK_SEARCH_HIGHLIGHT,
  createFindReplacePlugin: () => createFindReplacePlugin,
  decorateFindReplace: () => decorateFindReplace
});
module.exports = __toCommonJS(src_exports);

// src/createFindReplacePlugin.ts
var import_server2 = require("@udecode/plate-common/server");

// src/decorateFindReplace.ts
var import_server = require("@udecode/plate-common/server");
var decorateFindReplace = (editor, { key, type }) => ([node, path]) => {
  const ranges = [];
  const { search } = editor.pluginsByKey[key].options;
  if (!search || !(0, import_server.isText)(node)) {
    return ranges;
  }
  const { text } = node;
  const parts = text.toLowerCase().split(search.toLowerCase());
  let offset = 0;
  parts.forEach((part, i) => {
    if (i !== 0) {
      ranges.push({
        anchor: { offset: offset - search.length, path },
        focus: { offset, path },
        search,
        [type]: true
      });
    }
    offset = offset + part.length + search.length;
  });
  return ranges;
};

// src/createFindReplacePlugin.ts
var MARK_SEARCH_HIGHLIGHT = "search_highlight";
var createFindReplacePlugin = (0, import_server2.createPluginFactory)({
  decorate: decorateFindReplace,
  isLeaf: true,
  key: MARK_SEARCH_HIGHLIGHT
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MARK_SEARCH_HIGHLIGHT,
  createFindReplacePlugin,
  decorateFindReplace
});
