// @udecode/plate-slash-command/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ELEMENT_SLASH_INPUT: () => ELEMENT_SLASH_INPUT,
  KEY_SLASH_COMMAND: () => KEY_SLASH_COMMAND,
  createSlashPlugin: () => createSlashPlugin
});
module.exports = __toCommonJS(src_exports);

// src/createSlashPlugin.ts
var import_plate_combobox = require("@udecode/plate-combobox");
var import_server = require("@udecode/plate-common/server");
var KEY_SLASH_COMMAND = "slash_command";
var ELEMENT_SLASH_INPUT = "slash_input";
var createSlashPlugin = (0, import_server.createPluginFactory)({
  key: KEY_SLASH_COMMAND,
  options: {
    createComboboxInput: () => ({
      children: [{ text: "" }],
      type: ELEMENT_SLASH_INPUT
    }),
    trigger: "/",
    triggerPreviousCharPattern: /^\s?$/
  },
  plugins: [
    {
      isElement: true,
      isInline: true,
      isVoid: true,
      key: ELEMENT_SLASH_INPUT
    }
  ],
  withOverrides: import_plate_combobox.withTriggerCombobox
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ELEMENT_SLASH_INPUT,
  KEY_SLASH_COMMAND,
  createSlashPlugin
});
