// @udecode/plate-kbd/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  MARK_KBD: () => MARK_KBD,
  createKbdPlugin: () => createKbdPlugin
});
module.exports = __toCommonJS(src_exports);

// src/createKbdPlugin.ts
var import_server = require("@udecode/plate-common/server");
var MARK_KBD = "kbd";
var createKbdPlugin = (0, import_server.createPluginFactory)({
  deserializeHtml: {
    rules: [{ validNodeName: ["KBD"] }]
  },
  handlers: {
    onKeyDown: import_server.onKeyDownToggleMark
  },
  isLeaf: true,
  key: MARK_KBD
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MARK_KBD,
  createKbdPlugin
});
