// @udecode/plate-autoformat/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  KEY_AUTOFORMAT: () => KEY_AUTOFORMAT,
  autoformatArrow: () => autoformatArrow,
  autoformatBlock: () => autoformatBlock,
  autoformatComparison: () => autoformatComparison,
  autoformatDivision: () => autoformatDivision,
  autoformatEquality: () => autoformatEquality,
  autoformatFraction: () => autoformatFraction,
  autoformatLegal: () => autoformatLegal,
  autoformatLegalHtml: () => autoformatLegalHtml,
  autoformatMark: () => autoformatMark,
  autoformatMath: () => autoformatMath,
  autoformatOperation: () => autoformatOperation,
  autoformatPunctuation: () => autoformatPunctuation,
  autoformatSmartQuotes: () => autoformatSmartQuotes,
  autoformatSubscriptNumbers: () => autoformatSubscriptNumbers,
  autoformatSubscriptSymbols: () => autoformatSubscriptSymbols,
  autoformatSuperscriptNumbers: () => autoformatSuperscriptNumbers,
  autoformatSuperscriptSymbols: () => autoformatSuperscriptSymbols,
  autoformatText: () => autoformatText,
  createAutoformatPlugin: () => createAutoformatPlugin,
  getMatchPoints: () => getMatchPoints,
  getMatchRange: () => getMatchRange,
  isPreviousCharacterEmpty: () => isPreviousCharacterEmpty,
  onKeyDownAutoformat: () => onKeyDownAutoformat,
  withAutoformat: () => withAutoformat
});
module.exports = __toCommonJS(src_exports);

// src/createAutoformatPlugin.ts
var import_server8 = require("@udecode/plate-common/server");

// src/onKeyDownAutoformat.ts
var import_server = require("@udecode/plate-common/server");
var import_slate = require("slate");
var onKeyDownAutoformat = (editor, {
  options: { enableUndoOnDelete, rules }
}) => (e) => {
  if (e.defaultPrevented)
    return false;
  if (!(0, import_server.isHotkey)("backspace", { byKey: true }, e))
    return false;
  if (!rules)
    return false;
  if (!enableUndoOnDelete)
    return false;
  const { selection } = editor;
  if (!selection || !import_slate.Range.isCollapsed(selection))
    return;
  const [start, end] = import_slate.Range.edges(selection);
  const before = (0, import_server.getPointBefore)(editor, end, {
    distance: 1,
    unit: "character"
  });
  if (!start)
    return false;
  if (!before)
    return false;
  const charRange = { anchor: before, focus: start };
  if (!charRange)
    return false;
  const char = (0, import_server.getEditorString)(editor, charRange);
  if (!char)
    return false;
  const matchers = [...rules].filter((rule2) => {
    const textRule = rule2;
    if (textRule) {
      return textRule.mode === "text" && textRule.format === char;
    }
    return false;
  });
  if (!matchers || matchers.length === 0)
    return false;
  e.preventDefault();
  (0, import_server.deleteBackward)(editor, { unit: "character" });
  const rule = matchers[0];
  if (rule && typeof rule.match === "string") {
    (0, import_server.insertText)(editor, rule.match);
  } else {
    const matchArray = rule.match;
    if (matchArray && matchArray.length > 0) {
      (0, import_server.insertText)(editor, matchArray[0]);
    }
  }
  return true;
};

// src/withAutoformat.ts
var import_server7 = require("@udecode/plate-common/server");

// src/transforms/autoformatBlock.ts
var import_server2 = require("@udecode/plate-common/server");
var import_castArray2 = __toESM(require("lodash/castArray.js"));

// src/utils/getMatchRange.ts
var import_castArray = __toESM(require("lodash/castArray.js"));
var getMatchRange = ({
  match,
  trigger
}) => {
  let start;
  let end;
  if (typeof match === "object") {
    start = match.start;
    end = match.end;
  } else {
    start = match;
    end = start.split("").reverse().join("");
  }
  const triggers = trigger ? (0, import_castArray.default)(trigger) : [end.slice(-1)];
  end = trigger ? end : end.slice(0, -1);
  return {
    end,
    start,
    triggers
  };
};

// src/transforms/autoformatBlock.ts
var autoformatBlock = (editor, {
  allowSameTypeAbove = false,
  format,
  match: _match,
  matchByRegex = false,
  preFormat,
  text,
  trigger,
  triggerAtBlockStart = true,
  type = import_server2.ELEMENT_DEFAULT
}) => {
  const matches = (0, import_castArray2.default)(_match);
  for (const match of matches) {
    const { end, triggers } = getMatchRange({
      match: { end: match, start: "" },
      trigger
    });
    if (!triggers.includes(text))
      continue;
    let matchRange;
    if (triggerAtBlockStart) {
      matchRange = (0, import_server2.getRangeFromBlockStart)(editor);
      const hasVoidNode = (0, import_server2.someNode)(editor, {
        at: matchRange,
        match: (n) => (0, import_server2.isVoid)(editor, n)
      });
      if (hasVoidNode)
        continue;
      const textFromBlockStart = (0, import_server2.getEditorString)(editor, matchRange);
      const isMatched = matchByRegex ? !!textFromBlockStart.match(end) : end === textFromBlockStart;
      if (!isMatched)
        continue;
    } else {
      matchRange = (0, import_server2.getRangeBefore)(editor, editor.selection, {
        matchByRegex,
        matchString: end
      });
      if (!matchRange)
        continue;
    }
    if (!allowSameTypeAbove) {
      const isBelowSameBlockType = (0, import_server2.someNode)(editor, { match: { type } });
      if (isBelowSameBlockType)
        continue;
    }
    if (match.length > 1) {
      (0, import_server2.deleteText)(editor, {
        at: matchRange
      });
    }
    if (preFormat) {
      preFormat(editor);
    }
    if (format) {
      format(editor);
    } else {
      (0, import_server2.setElements)(
        editor,
        { type },
        {
          match: (n) => (0, import_server2.isBlock)(editor, n)
        }
      );
    }
    return true;
  }
  return false;
};

// src/transforms/autoformatMark.ts
var import_server5 = require("@udecode/plate-common/server");
var import_castArray3 = __toESM(require("lodash/castArray.js"));

// src/utils/getMatchPoints.ts
var import_server4 = require("@udecode/plate-common/server");

// src/utils/isPreviousCharacterEmpty.ts
var import_server3 = require("@udecode/plate-common/server");
var isPreviousCharacterEmpty = (editor, at) => {
  const range = (0, import_server3.getRangeBefore)(editor, at);
  if (range) {
    const text = (0, import_server3.getEditorString)(editor, range);
    if (text) {
      const noWhiteSpaceRegex = new RegExp(`\\S+`);
      return !text.match(noWhiteSpaceRegex);
    }
  }
  return true;
};

// src/utils/getMatchPoints.ts
var getMatchPoints = (editor, { end, start }) => {
  const selection = editor.selection;
  let beforeEndMatchPoint = selection.anchor;
  if (end) {
    beforeEndMatchPoint = (0, import_server4.getPointBeforeLocation)(editor, selection, {
      matchString: end
    });
    if (!beforeEndMatchPoint)
      return;
  }
  let afterStartMatchPoint;
  let beforeStartMatchPoint;
  if (start) {
    afterStartMatchPoint = (0, import_server4.getPointBeforeLocation)(editor, beforeEndMatchPoint, {
      afterMatch: true,
      matchString: start,
      skipInvalid: true
    });
    if (!afterStartMatchPoint)
      return;
    beforeStartMatchPoint = (0, import_server4.getPointBeforeLocation)(
      editor,
      beforeEndMatchPoint,
      {
        matchString: start,
        skipInvalid: true
      }
    );
    if (!isPreviousCharacterEmpty(editor, beforeStartMatchPoint))
      return;
  }
  return {
    afterStartMatchPoint,
    beforeEndMatchPoint,
    beforeStartMatchPoint
  };
};

// src/transforms/autoformatMark.ts
var autoformatMark = (editor, { ignoreTrim, match: _match, text, trigger, type }) => {
  if (!type)
    return false;
  const selection = editor.selection;
  const matches = (0, import_castArray3.default)(_match);
  for (const match of matches) {
    const { end, start, triggers } = getMatchRange({
      match,
      trigger
    });
    if (!triggers.includes(text))
      continue;
    const matched = getMatchPoints(editor, { end, start });
    if (!matched)
      continue;
    const { afterStartMatchPoint, beforeEndMatchPoint, beforeStartMatchPoint } = matched;
    const matchRange = {
      anchor: afterStartMatchPoint,
      focus: beforeEndMatchPoint
    };
    if (!ignoreTrim) {
      const matchText = (0, import_server5.getEditorString)(editor, matchRange);
      if (matchText.trim() !== matchText)
        continue;
    }
    if (end) {
      (0, import_server5.deleteText)(editor, {
        at: {
          anchor: beforeEndMatchPoint,
          focus: selection.anchor
        }
      });
    }
    const marks = (0, import_castArray3.default)(type);
    (0, import_server5.select)(editor, matchRange);
    marks.forEach((mark) => {
      editor.addMark(mark, true);
    });
    (0, import_server5.collapseSelection)(editor, { edge: "end" });
    (0, import_server5.removeMark)(editor, { key: marks, shouldChange: false });
    (0, import_server5.deleteText)(editor, {
      at: {
        anchor: beforeStartMatchPoint,
        focus: afterStartMatchPoint
      }
    });
    return true;
  }
  return false;
};

// src/transforms/autoformatText.ts
var import_server6 = require("@udecode/plate-common/server");
var import_castArray4 = __toESM(require("lodash/castArray.js"));
var autoformatText = (editor, { format, match: _match, text, trigger }) => {
  const selection = editor.selection;
  const matches = (0, import_castArray4.default)(_match);
  for (const match of matches) {
    const { end, start, triggers } = getMatchRange({
      match: Array.isArray(format) ? match : {
        end: match,
        start: ""
      },
      trigger
    });
    if (!triggers.includes(text))
      continue;
    const matched = getMatchPoints(editor, { end, start });
    if (!matched)
      continue;
    const { afterStartMatchPoint, beforeEndMatchPoint, beforeStartMatchPoint } = matched;
    if (end) {
      (0, import_server6.deleteText)(editor, {
        at: {
          anchor: beforeEndMatchPoint,
          focus: selection.anchor
        }
      });
    }
    if (typeof format === "function") {
      format(editor, matched);
    } else {
      const formatEnd = Array.isArray(format) ? format[1] : format;
      editor.insertText(formatEnd);
      if (beforeStartMatchPoint) {
        const formatStart = Array.isArray(format) ? format[0] : format;
        (0, import_server6.deleteText)(editor, {
          at: {
            anchor: beforeStartMatchPoint,
            focus: afterStartMatchPoint
          }
        });
        (0, import_server6.insertText)(editor, formatStart, {
          at: beforeStartMatchPoint
        });
      }
    }
    return true;
  }
  return false;
};

// src/withAutoformat.ts
var withAutoformat = (editor, { options: { rules } }) => {
  const { insertText: insertText3 } = editor;
  editor.insertText = (text) => {
    var _a;
    if (!(0, import_server7.isCollapsed)(editor.selection))
      return insertText3(text);
    for (const rule of rules) {
      const { insertTrigger, mode = "text", query } = rule;
      if (query && !query(editor, __spreadProps(__spreadValues({}, rule), { text })))
        continue;
      const autoformatter = {
        block: autoformatBlock,
        mark: autoformatMark,
        text: autoformatText
      };
      if ((_a = autoformatter[mode]) == null ? void 0 : _a.call(autoformatter, editor, __spreadProps(__spreadValues({}, rule), {
        text
      }))) {
        return insertTrigger && insertText3(text);
      }
    }
    insertText3(text);
  };
  return editor;
};

// src/createAutoformatPlugin.ts
var KEY_AUTOFORMAT = "autoformat";
var createAutoformatPlugin = (0, import_server8.createPluginFactory)({
  handlers: {
    onKeyDown: onKeyDownAutoformat
  },
  key: KEY_AUTOFORMAT,
  options: {
    rules: []
  },
  withOverrides: withAutoformat
});

// src/rules/autoformatArrow.ts
var autoformatArrow = [
  {
    format: "\u2192",
    match: "->",
    mode: "text"
  },
  {
    format: "\u2190",
    match: "<-",
    mode: "text"
  },
  {
    format: "\u21D2",
    match: "=>",
    mode: "text"
  },
  {
    format: "\u21D0",
    match: ["<=", "\u2264="],
    mode: "text"
  }
];

// src/rules/autoformatLegal.ts
var autoformatLegal = [
  {
    format: "\u2122",
    match: ["(tm)", "(TM)"],
    mode: "text"
  },
  {
    format: "\xAE",
    match: ["(r)", "(R)"],
    mode: "text"
  },
  {
    format: "\xA9",
    match: ["(c)", "(C)"],
    mode: "text"
  }
];
var autoformatLegalHtml = [
  {
    format: "\u2122",
    match: "&trade;",
    mode: "text"
  },
  {
    format: "\xAE",
    match: "&reg;",
    mode: "text"
  },
  {
    format: "\xA9",
    match: "&copy;",
    mode: "text"
  },
  {
    format: "\xA7",
    match: "&sect;",
    mode: "text"
  }
];

// src/rules/autoformatPunctuation.ts
var autoformatPunctuation = [
  {
    format: "\u2014",
    match: "--",
    mode: "text"
  },
  {
    format: "\u2026",
    match: "...",
    mode: "text"
  },
  {
    format: "\xBB",
    match: ">>",
    mode: "text"
  },
  {
    format: "\xAB",
    match: "<<",
    mode: "text"
  }
];

// src/rules/autoformatSmartQuotes.ts
var autoformatSmartQuotes = [
  {
    format: ["\u201C", "\u201D"],
    match: '"',
    mode: "text"
  },
  {
    format: ["\u2018", "\u2019"],
    match: "'",
    mode: "text"
  }
];

// src/rules/math/autoformatComparison.ts
var autoformatComparison = [
  {
    format: "\u226F",
    match: "!>",
    mode: "text"
  },
  {
    format: "\u226E",
    match: "!<",
    mode: "text"
  },
  {
    format: "\u2265",
    match: ">=",
    mode: "text"
  },
  {
    format: "\u2264",
    match: "<=",
    mode: "text"
  },
  {
    format: "\u2271",
    match: "!>=",
    mode: "text"
  },
  {
    format: "\u2270",
    match: "!<=",
    mode: "text"
  }
];

// src/rules/math/autoformatEquality.ts
var autoformatEquality = [
  {
    format: "\u2260",
    match: "!=",
    mode: "text"
  },
  {
    format: "\u2261",
    match: "==",
    mode: "text"
  },
  {
    format: "\u2262",
    match: ["!==", "\u2260="],
    mode: "text"
  },
  {
    format: "\u2248",
    match: "~=",
    mode: "text"
  },
  {
    format: "\u2249",
    match: "!~=",
    mode: "text"
  }
];

// src/rules/math/autoformatFraction.ts
var autoformatFraction = [
  {
    format: "\xBD",
    match: "1/2",
    mode: "text"
  },
  {
    format: "\u2153",
    match: "1/3",
    mode: "text"
  },
  {
    format: "\xBC",
    match: "1/4",
    mode: "text"
  },
  {
    format: "\u2155",
    match: "1/5",
    mode: "text"
  },
  {
    format: "\u2159",
    match: "1/6",
    mode: "text"
  },
  {
    format: "\u2150",
    match: "1/7",
    mode: "text"
  },
  {
    format: "\u215B",
    match: "1/8",
    mode: "text"
  },
  {
    format: "\u2151",
    match: "1/9",
    mode: "text"
  },
  {
    format: "\u2152",
    match: "1/10",
    mode: "text"
  },
  {
    format: "\u2154",
    match: "2/3",
    mode: "text"
  },
  {
    format: "\u2156",
    match: "2/5",
    mode: "text"
  },
  {
    format: "\xBE",
    match: "3/4",
    mode: "text"
  },
  {
    format: "\u2157",
    match: "3/5",
    mode: "text"
  },
  {
    format: "\u215C",
    match: "3/8",
    mode: "text"
  },
  {
    format: "\u2158",
    match: "4/5",
    mode: "text"
  },
  {
    format: "\u215A",
    match: "5/6",
    mode: "text"
  },
  {
    format: "\u215D",
    match: "5/8",
    mode: "text"
  },
  {
    format: "\u215E",
    match: "7/8",
    mode: "text"
  }
];

// src/rules/math/autoformatOperation.ts
var autoformatDivision = [
  {
    format: "\xF7",
    match: "//",
    mode: "text"
  }
];
var autoformatOperation = [
  {
    format: "\xB1",
    match: "+-",
    mode: "text"
  },
  {
    format: "\u2030",
    match: "%%",
    mode: "text"
  },
  {
    format: "\u2031",
    match: ["%%%", "\u2030%"],
    mode: "text"
  },
  ...autoformatDivision
];

// src/rules/math/autoformatSubscript.ts
var autoformatSubscriptNumbers = [
  {
    format: "\u2080",
    match: "~0",
    mode: "text"
  },
  {
    format: "\u2081",
    match: "~1",
    mode: "text"
  },
  {
    format: "\u2082",
    match: "~2",
    mode: "text"
  },
  {
    format: "\u2083",
    match: "~3",
    mode: "text"
  },
  {
    format: "\u2084",
    match: "~4",
    mode: "text"
  },
  {
    format: "\u2085",
    match: "~5",
    mode: "text"
  },
  {
    format: "\u2086",
    match: "~6",
    mode: "text"
  },
  {
    format: "\u2087",
    match: "~7",
    mode: "text"
  },
  {
    format: "\u2088",
    match: "~8",
    mode: "text"
  },
  {
    format: "\u2089",
    match: "~9",
    mode: "text"
  }
];
var autoformatSubscriptSymbols = [
  {
    format: "\u208A",
    match: "~+",
    mode: "text"
  },
  {
    format: "\u208B",
    match: "~-",
    mode: "text"
  }
];

// src/rules/math/autoformatSuperscript.ts
var autoformatSuperscriptNumbers = [
  {
    format: "\u2070",
    match: "^0",
    mode: "text"
  },
  {
    format: "\xB9",
    match: "^1",
    mode: "text"
  },
  {
    format: "\xB2",
    match: "^2",
    mode: "text"
  },
  {
    format: "\xB3",
    match: "^3",
    mode: "text"
  },
  {
    format: "\u2074",
    match: "^4",
    mode: "text"
  },
  {
    format: "\u2075",
    match: "^5",
    mode: "text"
  },
  {
    format: "\u2076",
    match: "^6",
    mode: "text"
  },
  {
    format: "\u2077",
    match: "^7",
    mode: "text"
  },
  {
    format: "\u2078",
    match: "^8",
    mode: "text"
  },
  {
    format: "\u2079",
    match: "^9",
    mode: "text"
  }
];
var autoformatSuperscriptSymbols = [
  {
    format: "\xB0",
    match: "^o",
    mode: "text"
  },
  {
    format: "\u207A",
    match: "^+",
    mode: "text"
  },
  {
    format: "\u207B",
    match: "^-",
    mode: "text"
  }
];

// src/rules/math/autoformatMath.ts
var autoformatMath = [
  ...autoformatComparison,
  ...autoformatEquality,
  ...autoformatOperation,
  ...autoformatFraction,
  ...autoformatSuperscriptSymbols,
  ...autoformatSubscriptSymbols,
  ...autoformatSuperscriptNumbers,
  ...autoformatSubscriptNumbers
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KEY_AUTOFORMAT,
  autoformatArrow,
  autoformatBlock,
  autoformatComparison,
  autoformatDivision,
  autoformatEquality,
  autoformatFraction,
  autoformatLegal,
  autoformatLegalHtml,
  autoformatMark,
  autoformatMath,
  autoformatOperation,
  autoformatPunctuation,
  autoformatSmartQuotes,
  autoformatSubscriptNumbers,
  autoformatSubscriptSymbols,
  autoformatSuperscriptNumbers,
  autoformatSuperscriptSymbols,
  autoformatText,
  createAutoformatPlugin,
  getMatchPoints,
  getMatchRange,
  isPreviousCharacterEmpty,
  onKeyDownAutoformat,
  withAutoformat
});
