// @udecode/plate-serializer-csv/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  KEY_DESERIALIZE_CSV: () => KEY_DESERIALIZE_CSV,
  createDeserializeCsvPlugin: () => createDeserializeCsvPlugin,
  deserializeCsv: () => deserializeCsv
});
module.exports = __toCommonJS(src_exports);

// src/deserializer/createDeserializeCsvPlugin.ts
var import_server2 = require("@udecode/plate-common/server");

// src/deserializer/utils/deserializeCsv.ts
var import_server = require("@udecode/plate-common/server");
var import_papaparse = __toESM(require("papaparse"));
var { parse } = import_papaparse.default;
var isValidCsv = (data, errors, errorTolerance) => {
  if (errorTolerance < 0)
    errorTolerance = 0;
  return !(!data || data.length < 2 || data[0].length < 2 || data[1].length < 2 || errors.length > 0 && errors.length > errorTolerance * data.length);
};
var deserializeCsv = (editor, _a) => {
  var _b = _a, {
    data
  } = _b, parseOptions = __objRest(_b, [
    "data"
  ]);
  const {
    options: { errorTolerance, parseOptions: pluginParseOptions }
  } = (0, import_server.getPlugin)(
    editor,
    KEY_DESERIALIZE_CSV
  );
  const testCsv = parse(data, { preview: 2 });
  if (testCsv.errors.length === 0) {
    const csv = parse(data, __spreadValues(__spreadValues({}, pluginParseOptions), parseOptions));
    if (!isValidCsv(
      csv.data,
      csv.errors,
      errorTolerance
    ))
      return;
    const paragraph = (0, import_server.getPluginType)(editor, import_server.ELEMENT_DEFAULT);
    const table = (0, import_server.getPluginType)(editor, "table");
    const th = (0, import_server.getPluginType)(editor, "th");
    const tr = (0, import_server.getPluginType)(editor, "tr");
    const td = (0, import_server.getPluginType)(editor, "td");
    const ast = {
      children: [],
      type: table
    };
    if (csv.meta.fields) {
      ast.children.push({
        children: csv.meta.fields.map((field) => ({
          children: [{ children: [{ text: field }], type: paragraph }],
          type: th
        })),
        type: tr
      });
      for (const row of csv.data) {
        ast.children.push({
          children: csv.meta.fields.map((field) => ({
            children: [
              { children: [{ text: row[field] || "" }], type: paragraph }
            ],
            type: td
          })),
          type: tr
        });
      }
    } else {
      for (const row of csv.data) {
        ast.children.push({
          children: [],
          type: tr
        });
        for (const cell of row) {
          ast.children.at(-1).children.push({
            children: [{ children: [{ text: cell }], type: paragraph }],
            type: td
          });
        }
      }
    }
    return [
      {
        children: [{ text: "" }],
        type: paragraph
      },
      ast,
      {
        children: [{ text: "" }],
        type: paragraph
      }
    ];
  }
};

// src/deserializer/createDeserializeCsvPlugin.ts
var KEY_DESERIALIZE_CSV = "deserializeCsv";
var createDeserializeCsvPlugin = (0, import_server2.createPluginFactory)({
  key: KEY_DESERIALIZE_CSV,
  options: {
    errorTolerance: 0.25,
    parseOptions: {
      header: true
    }
  },
  then: (editor) => ({
    editor: {
      insertData: {
        format: "text/plain",
        getFragment: ({ data }) => deserializeCsv(editor, { data })
      }
    }
  })
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KEY_DESERIALIZE_CSV,
  createDeserializeCsvPlugin,
  deserializeCsv
});
