// @udecode/plate-utils/dist/server.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/server.ts
var server_exports = {};
__export(server_exports, {
  PlateElement: () => PlateElement,
  PlateLeaf: () => PlateLeaf,
  addSelectedRow: () => addSelectedRow,
  blockSelectedIds: () => blockSelectedIds,
  createNodeHOC: () => createNodeHOC,
  createNodesHOC: () => createNodesHOC,
  createNodesWithHOC: () => createNodesWithHOC,
  defaultsDeepToNodes: () => defaultsDeepToNodes,
  hasBlockSelected: () => hasBlockSelected,
  isBlockSelected: () => isBlockSelected,
  isType: () => isType,
  moveSelectionByOffset: () => moveSelectionByOffset,
  onKeyDownToggleElement: () => onKeyDownToggleElement,
  onKeyDownToggleMark: () => onKeyDownToggleMark,
  usePlateElement: () => usePlateElement,
  usePlateLeaf: () => usePlateLeaf
});
module.exports = __toCommonJS(server_exports);

// src/shared/PlateElement.tsx
var import_react = __toESM(require("react"));
var import_react_utils = require("@udecode/react-utils");
var import_clsx = require("clsx");
var usePlateElement = (props) => {
  const _a = props, {
    attributes,
    editor,
    element,
    elementToAttributes,
    nodeProps
  } = _a, rootProps = __objRest(_a, [
    "attributes",
    "editor",
    "element",
    "elementToAttributes",
    "nodeProps"
  ]);
  return {
    props: __spreadProps(__spreadValues(__spreadValues(__spreadValues(__spreadValues({}, attributes), rootProps), nodeProps), elementToAttributes == null ? void 0 : elementToAttributes(element)), {
      className: (0, import_clsx.clsx)(props.className, nodeProps == null ? void 0 : nodeProps.className)
    }),
    ref: (0, import_react_utils.useComposedRef)(props.ref, attributes.ref)
  };
};
var PlateElement = import_react.default.forwardRef(
  (props, ref) => {
    const { props: rootProps, ref: rootRef } = usePlateElement(__spreadProps(__spreadValues({}, props), {
      ref
    }));
    return /* @__PURE__ */ import_react.default.createElement(import_react_utils.Box, __spreadProps(__spreadValues({}, rootProps), { ref: rootRef }));
  }
);
PlateElement.displayName = "PlateElement";

// src/shared/PlateLeaf.tsx
var import_react2 = __toESM(require("react"));
var import_react_utils2 = require("@udecode/react-utils");
var import_clsx2 = require("clsx");
var usePlateLeaf = (props) => {
  const _a = props, {
    attributes,
    editor,
    leaf,
    leafToAttributes,
    nodeProps,
    text
  } = _a, rootProps = __objRest(_a, [
    "attributes",
    "editor",
    "leaf",
    "leafToAttributes",
    "nodeProps",
    "text"
  ]);
  return {
    props: __spreadProps(__spreadValues(__spreadValues(__spreadValues(__spreadValues({}, attributes), rootProps), nodeProps), leafToAttributes == null ? void 0 : leafToAttributes(leaf)), {
      className: (0, import_clsx2.clsx)(props.className, nodeProps == null ? void 0 : nodeProps.className)
    }),
    ref: (0, import_react_utils2.useComposedRef)(props.ref, attributes.ref)
  };
};
var PlateLeaf = import_react2.default.forwardRef(
  (props, ref) => {
    const { props: rootProps, ref: rootRef } = usePlateLeaf(__spreadProps(__spreadValues({}, props), { ref }));
    return /* @__PURE__ */ import_react2.default.createElement(import_react_utils2.Text, __spreadProps(__spreadValues({}, rootProps), { ref: rootRef }));
  }
);
PlateLeaf.displayName = "PlateLeaf";

// src/shared/addSelectedRow.ts
var addSelectedRow = (editor, id, options = {}) => {
  const { blockSelectionStore } = editor;
  if (blockSelectionStore) {
    blockSelectionStore.set.addSelectedRow(id, options);
  }
};

// src/shared/blockSelectedIds.ts
var blockSelectedIds = (editor) => {
  const { blockSelectionStore } = editor;
  if (blockSelectionStore) {
    return blockSelectionStore.get.selectedIds();
  }
};

// src/shared/createNodeHOC.tsx
var import_react3 = __toESM(require("react"));
var createNodeHOC = (HOC) => (Component, props) => function hoc(childrenProps) {
  return /* @__PURE__ */ import_react3.default.createElement(HOC, __spreadValues({}, __spreadValues(__spreadValues({}, childrenProps), props)), /* @__PURE__ */ import_react3.default.createElement(Component, __spreadValues({}, childrenProps)));
};

// src/shared/createNodesHOC.tsx
var import_castArray = __toESM(require("lodash/castArray.js"));
var import_merge = __toESM(require("lodash/merge.js"));
var createHOC = (withHOC) => {
  return (components, options) => {
    const _components = __spreadValues({}, components);
    const optionsByKey = {};
    const optionsList = (0, import_castArray.default)(options);
    optionsList.forEach((_a) => {
      var _b = _a, { key, keys } = _b, opt = __objRest(_b, ["key", "keys"]);
      const _keys = key ? [key] : keys != null ? keys : Object.keys(_components);
      _keys.forEach((_key) => {
        optionsByKey[_key] = (0, import_merge.default)(optionsByKey[_key], opt);
      });
    });
    Object.keys(optionsByKey).forEach((key) => {
      if (!_components[key])
        return;
      _components[key] = withHOC(_components[key], optionsByKey[key]);
    });
    return _components;
  };
};
var createNodesHOC = (HOC) => {
  return createHOC(createNodeHOC(HOC));
};
var createNodesWithHOC = (withHOC) => {
  return createHOC(withHOC);
};

// src/shared/defaultsDeepToNodes.ts
var import_server = require("@udecode/plate-core/server");
var import_defaults = __toESM(require("lodash/defaults.js"));
var defaultsDeepToNodes = (options) => {
  (0, import_server.applyDeepToNodes)(__spreadProps(__spreadValues({}, options), { apply: import_defaults.default }));
};

// src/shared/hasBlockSelected.ts
var hasBlockSelected = (editor) => {
  const ids = blockSelectedIds(editor);
  return ids && ids.size > 0;
};

// src/shared/isBlockSelected.ts
var isBlockSelected = (editor, id) => {
  const ids = blockSelectedIds(editor);
  return ids == null ? void 0 : ids.has(id);
};

// src/shared/isType.ts
var import_server2 = require("@udecode/plate-core/server");
var import_castArray2 = __toESM(require("lodash/castArray.js"));
var isType = (editor, node, key) => {
  const keys = (0, import_castArray2.default)(key);
  const types = [];
  keys.forEach((_key) => types.push((0, import_server2.getPluginType)(editor, _key)));
  return types.includes(node == null ? void 0 : node.type);
};

// src/shared/moveSelectionByOffset.ts
var import_server3 = require("@udecode/plate-core/server");
var import_slate = require("@udecode/slate");
var import_slate2 = require("slate");
var moveSelectionByOffset = (editor, { query = () => true } = {}) => (event) => {
  const { selection } = editor;
  if (!selection || import_slate2.Range.isExpanded(selection) || !query(editor)) {
    return false;
  }
  if ((0, import_server3.isHotkey)("left", event)) {
    event.preventDefault();
    (0, import_slate.moveSelection)(editor, { reverse: true, unit: "offset" });
    return true;
  }
  if ((0, import_server3.isHotkey)("right", event)) {
    event.preventDefault();
    (0, import_slate.moveSelection)(editor, { unit: "offset" });
    return true;
  }
};

// src/shared/onKeyDownToggleElement.ts
var import_server4 = require("@udecode/plate-core/server");
var import_castArray3 = __toESM(require("lodash/castArray.js"));
var onKeyDownToggleElement = (editor, { options: { hotkey }, type }) => (e) => {
  if (e.defaultPrevented)
    return;
  const defaultType = (0, import_server4.getPluginType)(editor, import_server4.ELEMENT_DEFAULT);
  if (!hotkey)
    return;
  const hotkeys = (0, import_castArray3.default)(hotkey);
  for (const _hotkey of hotkeys) {
    if ((0, import_server4.isHotkey)(_hotkey, e)) {
      e.preventDefault();
      (0, import_server4.toggleNodeType)(editor, {
        activeType: type,
        inactiveType: defaultType
      });
      return;
    }
  }
};

// src/shared/onKeyDownToggleMark.ts
var import_server5 = require("@udecode/plate-core/server");
var import_slate_utils = require("@udecode/slate-utils");
var onKeyDownToggleMark = (editor, {
  options: { clear, hotkey },
  type
}) => (e) => {
  if (e.defaultPrevented)
    return;
  if (!hotkey)
    return;
  if ((0, import_server5.isHotkey)(hotkey, e)) {
    e.preventDefault();
    (0, import_slate_utils.toggleMark)(editor, { clear, key: type });
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PlateElement,
  PlateLeaf,
  addSelectedRow,
  blockSelectedIds,
  createNodeHOC,
  createNodesHOC,
  createNodesWithHOC,
  defaultsDeepToNodes,
  hasBlockSelected,
  isBlockSelected,
  isType,
  moveSelectionByOffset,
  onKeyDownToggleElement,
  onKeyDownToggleMark,
  usePlateElement,
  usePlateLeaf
});
